### PR TITLE
Plaquette design

### PR DESCRIPTION
**Pull request by nelimee**
Opening the pull request for code review.

### CODE REVIEW

**Comment by smburdick** [comment]
I'd be happy to review this, Adrien.

**Comment by nelimee** [comment]
> @nelimee I'd be happy to review this, Adrien.

Thanks! If you have any question, feel free to send me a mail (the best way to reach me quickly), we can schedule a quick virtual meeting if we are both available. 

You can also comment here, but I will likely be less reactive than a direct mail.

File: python_prototype/main.py
Comment (smburdick) [inline python_prototype/main.py:111]: Running `python main.py` produces this error repeatedly
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 289, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspaces/tqec/python_prototype/main.py", line 122, in <module>
    collected_surface_code_stats: list[sinter.TaskStats] = sinter.collect(
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection.py", line 380, in collect
    for progress in iter_collect(
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection.py", line 191, in iter_collect
    manager.start_workers(num_workers)
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection_work_manager.py", line 85, in start_workers
    w.start()
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

File: python_prototype/src/tqec/constructions/qubit.py
Comment (smburdick) [inline python_prototype/src/tqec/constructions/qubit.py:26]: What do the different types represent? An enum may be useful for them

File: python_prototype/src/tqec/enums.py
Comment (smburdick) [inline python_prototype/src/tqec/enums.py:6]: Why not use a network graph to represent the grid of qubits? That way, each qubit doesn't need to know its position of relative position to others -- it is encoded by the edges of the graph. There may be reasons for doing it this way, that are unclear to me though.

**Review by smburdick — APPROVED** [review]
Approving, with some usability comments, for the purpose of merging the prototype into the codebase.

Looks like a lot of work is done here, and I'm looking forward to diving deeper into the specifics. As we find issues, we can add them to the issue folder, and work on them individually in line with the established tasking.

In the future -- and this goes for all of us -- be sure to submit PRs containing smaller chunks of code; this one is too long to navigate comfortably.

File: python_prototype/README.md
Comment (nelimee) [reply python_prototype/README.md:54]: Should be here: https://github.com/QCHackers/tqec/blob/789ae72a22efa64904a661c7ab65769deea22d77/python_prototype/src/tqec/templates/orchestrator.py#L55

File: python_prototype/TODO.md
Comment (nelimee) [reply python_prototype/TODO.md:1]: Agreed, but this file is not up-to-date anymore.

File: python_prototype/main.py
Comment (nelimee) [reply python_prototype/main.py:111]: Yes, I got that too. Seems like a `sinter` error. It works on notebooks, so I ended up using notebooks because solving that was not the top priority.

File: python_prototype/src/tqec/constructions/qubit.py
Comment (nelimee) [reply python_prototype/src/tqec/constructions/qubit.py:26]: You mean `ScalableAlternatingSquare` versus `ScalableRectangle`? These are lacking documentation, but I do not see how an enumeration would be useful. Could you elaborate?

File: python_prototype/TODO.md
Comment (smburdick) [reply python_prototype/TODO.md:1]: Then it should be removed.

File: python_prototype/main.py
Comment (smburdick) [reply python_prototype/main.py:111]: Good point, I saw the option to review the notebooks from this PR. Would you like a review from there as well?

File: python_prototype/src/tqec/enums.py
Comment (nelimee) [reply python_prototype/src/tqec/enums.py:6]: The graph approach is more generic and may be simpler. On this specific enumeration, the "coordinates" are not really qubits, they are mostly multiplying factors that helps in computing the position of different corners in
https://github.com/QCHackers/tqec/blob/789ae72a22efa64904a661c7ab65769deea22d77/python_prototype/src/tqec/templates/orchestrator.py#L45-L48

File: python_prototype/TODO.md
Comment (nelimee) [reply python_prototype/TODO.md:1]: Will do that

File: python_prototype/src/tqec/constructions/qubit.py
Comment (smburdick) [reply python_prototype/src/tqec/constructions/qubit.py:26]: In these lines you're plugging in numbers corresponding to the types. It's unclear what these are -- can you explain?

Usually we like to avoid magical numbers when possible, and an enum is one way to do that

File: python_prototype/main.py
Comment (nelimee) [reply python_prototype/main.py:111]: I am not the one that configured that. I do not use the ReviewNB that seems to be a third-party service. So I would lean on the "no I do not need you to review on that website" side. If it is simpler for you though, you can do so.

File: python_prototype/src/tqec/constructions/qubit.py
Comment (nelimee) [reply python_prototype/src/tqec/constructions/qubit.py:26]: You mean the `[3, 4]` and `[0, 1]`? These are plaquette indices, the numbers that will end up appearing in the final 2-dimensional array representing the error-correction code. I tried to make that explicit enough by introducing the `TemplateWithIndices` class, to avoid a basic `tuple` and to have named attributes that can be used to document the meaning of each field. This is not sufficient?

Comment (smburdick) [reply python_prototype/src/tqec/constructions/qubit.py:26]: I think that is fine. Thanks for the explanation.

File: python_prototype/main.py
Comment (smburdick) [reply python_prototype/main.py:111]: Are the notebooks meant for tutorial purposes only? If so, I don't think they need to be rigorously reviewed.

Comment (nelimee) [reply python_prototype/main.py:111]: Either documentation or as an easy way of testing the results obtained by the code. They can be thought as outside the library, so they can effectively be excluded from review in my opinion.

File: python_prototype/src/tqec/enums.py
Comment (smburdick) [reply python_prototype/src/tqec/enums.py]: We can use this. You are right, using a graph library may compound the complexity.

I'd be curious to know how significant the coordinate system is to this project as it grows -- it may be worthwhile to support multiple, equivalent ones, but that may only need to be reflected on the frontend.

**Comment by nelimee** [comment]
The commits I just pushed are, in my opinion, the last commits of this branch. From this point on, I think that the code is mature enough to be merged and iterated on using multiple, issue-based, and **smaller** (sorry for this one) pull-requests.
I'll wait for an approval before merging. If you want to merge directly, go ahead.

**Comment by rryoung98** [comment]
> The commits I just pushed are, in my opinion, the last commits of this branch. From this point on, I think that the code is mature enough to be merged and iterated on using multiple, issue-based, and **smaller** (sorry for this one) pull-requests. I'll wait for an approval before merging. If you want to merge directly, go ahead.

Thanks! Will merge today seems like the pip install is working now.

**Comment by nelimee** [comment]
Nice! I'll merge now then, so that I can open new branches and document the things that need to be done with issues in the afternoon.

**Pull request merged**
